### PR TITLE
[#11] Unduplicated errors

### LINKED ISSUE #11: Usability: Duplicated errors shouldn't be reported
# Why?

A degenerate example of us using List for error reporting would be the following parser application:

```
p = string "123"
Parsec.run "-" (p <|> p <|> p) "322"
````

The reference implementation of megaparsec is using `Set`s for error reporting, thus will report "Failed to parse line 1 symbol 1: found '3, expected '1". Our parser will report: "Failed to parse line 1 symbol 1: found '3, expected '1. Failed to parse line 1 symbol 1: found '3, expected '1. Failed to parse line 1 symbol 1: found '3, expected '1.".

# How?

 - [ ] Figure out how to append to HashSets.
 - [ ] Validate that all the types can append to HashSets.
 - [ ] Switch Lists to HashSets and append in errors.

### PR DESCRIPTION
**Pull request by ixahedron**
Problem: currently we use a list to store suggestions/hints/expected values for our parsing errors. This is simple, yet can lead to nonsensical duplicate errors. Example:

```lean
  let p  : Parsec Char String Unit String := string "yatima"
  let _ ← parseTestP (p <|> p <|> p) "Yatima"
```

results in:

```
  1:1:
  |
1 | Yatima
  | ^^^^^^
unexpected  "Yatima"
expecting  "yatima", "yatima", or "yatima"
```

Solution: switched to using RBMap instead of List.

Closes #11

### CODE REVIEW

File: Megaparsec/Errors.lean [hunk 1]
```diff
@@ -58,6 +69,36 @@ inductive ErrorFancy (E : Type u) where
...
+  | .eq, .eq => .eq
+  | .eq, .lt => .gt
+  | .lt, .lt => .eq
+  | .lt,   _ => .lt
+
+private def compareIndents : (Ordering × Pos × Pos) → (Ordering × Pos × Pos)
+                           → Ordering
+  | (o1, f1, u1), (o2, f2, u2) =>
+    match compare u1 u2, compare f1 f2 with
+                  -- we compare orderings as that's the last differing parameter
```
Comment (cognivore) [inline Megaparsec/Errors.lean:85]: I still don't understand it. Can you remind me the design of indent triples (as in, what does the first element in the triple mean?).

I understand the logic but don't understand the semantics.

File: Megaparsec/Errors/Bundle.lean [hunk 2]
```diff
@@ -27,7 +27,7 @@ private def makePEBfs [Printable β] [ToString E] [Streamable ℘] : ((String 
           | .trivial _     none _ => 1
           | .trivial _ (some x) _ => errorItemLength x
           | .fancy   _         xs =>
-            xs.foldl (fun acc e => max acc (errorFancyLength e)) 1
+            xs.fold (fun acc e _ => max acc (errorFancyLength e)) 1
```
Comment (cognivore) [inline Megaparsec/Errors/Bundle.lean:30]: What's the difference between what we had before and this? I didn't think about it too much, I admit, but I'd like to know why this change is necessary.

File: Megaparsec/Errors/ParseError.lean [hunk 3]
```diff
@@ -45,6 +51,20 @@ def messageItemsPretty (pref : String) (ts : List String) : String :=
...
+def RBMap.unitMap (xs : List α) : RBMap α Unit cmp :=
+  RBMap.ofList $ xs.map (fun x => (x, ()))
+
+/-
+  Merge two RBMaps, always taking the first value in case of a key
+  being present in both maps. Intended for set simulation.
+-/
+def RBMap.mergeKeySets (m1 : RBMap α β cmp) m2 := RBMap.mergeBy (fun _ v _ => v) m1 m2
+
+instance : Append (RBMap α Unit cmp) := ⟨RBMap.mergeKeySets⟩
```
Comment (cognivore) [inline Megaparsec/Errors/ParseError.lean:66]: Smart!

File: Megaparsec/MonadParsec.lean [hunk 4]
```diff
@@ -36,7 +36,7 @@ On type variables:
 - `γ` is the type of the thing we're parsing out of the source.
 
 -/
-class MonadParsec (m : Type u → Type v) (℘ α E β : Type u) where
+class MonadParsec (m : Type u → Type v) (℘ α E β : Type u) [Ord β] [Ord E] where
```
Comment (cognivore) [inline Megaparsec/MonadParsec.lean:39]: I disagree that these constraints belong here. Our typeclass is jackal as it is, and adding more hoops is kind of bad.

We can keep the constraint in the instances we provide and let the users come up with their own versions of deduplication, if they choose to accept that challenge.

(Sidenote: I still maintain that the pipe dream of extending MonadParsec by third-party users isn't much more than an overengineering. As I said in my analysis, if this polymorphism would've been actually needed, we wouldn't have had several different libraries defining their own typeclass in Haskell ecosystem, but rather all the Haskell parsecs would have been libraries depending on the original parsec).

File: Megaparsec/Ok.lean [hunk 5]
```diff
@@ -6,13 +6,13 @@ open Megaparsec.ParserState
...
-def Ok (m : Type f → Type v) (β ℘ E γ : Type u) (ξ : Type f) :=
+def Ok (m : Type f → Type v) (β ℘ E γ : Type u) (ξ : Type f) [Ord β] [Ord E] :=
   (γ → State β ℘ E → Hints β → m ξ)
 
 /-
 `accHints hs c` adds hints to “OK” continuation that will add given
 hints @hs@ to third argument of original continuation c
 -/
-def accHints (hs : Hints β)
+def accHints [Ord β] [Ord E] (hs : Hints β)
```
Comment (cognivore) [inline Megaparsec/Ok.lean:16]: In some places you have constraints in the end of argument list, in some -- in the beginning. Can you explain this decision or is it arbitrary?

**Review by cognivore — CHANGES_REQUESTED** [review]
I strongly advise dropping all the constraints from most of typeclasses and instead putting those in the instances. I also requested some comments.

When that is sorted (I think, it's a simple code deletion), I approve.

**Comment by cognivore** [comment]
In general, I don't think it's a big deal to ask for `Ord`. No matter which deduplication method we'd choose, we'd still need some sort of constraint on $$ \beta $$ and $$ E $$. Why not `Ord`. 

I disagree, however, that exactly `Ord` is necessary for this. That's one of the reasons I propose you to push `Ord` constraint down to instances land, rather than keeping it in the typeclass land.

Overall, I really like the changes. Great job! 🙇 

PS
I didn't run the code, please make sure that the "test" suite works before merging. :)

File: Megaparsec/Errors.lean (hunk 1 above)
Comment (ixahedron) [reply Megaparsec/Errors.lean:85]: It's mostly for formatting AFAICT, or rather for better location pointing. The indent triple is `fromPos`, `uptoPos`, and this ordering, and it's used in the `ToString` instance to say whether the correct indentation should be greater than `fromPos` or equal or less. So it's IMO the least important parameter.

File: Megaparsec/Errors/Bundle.lean (hunk 2 above)
Comment (ixahedron) [reply Megaparsec/Errors/Bundle.lean:30]: `RBMap` doesn't have `foldl`, and `xs` was changed from `List` to `RBMap`.

File: Megaparsec/Errors.lean (hunk 1 above)
Comment (cognivore) [reply Megaparsec/Errors.lean:103]: I still don't understand it, sorry. I think it warrants a docstring, explaining what happens if the indentation is bigger, what if less than, what if equal.

File: Megaparsec/Errors/Bundle.lean (hunk 2 above)
Comment (cognivore) [reply Megaparsec/Errors/Bundle.lean:30]: 🤦 thanks for explaining! I should have caught it during the review.

**Review by cognivore — APPROVED** [review]

**Comment by cognivore** [comment]
Approved changes! Feel free to merge 🙇
